mw_basic_auth.go

```diff
@@ -117,6 +106,67 @@ func (k *BasicAuthKeyIsValid) ProcessRequest(w http.ResponseWriter, r *http.Requ
 		ctxSetAuthToken(r, keyName)
 	}
 
-	// Request is valid, carry on
 	return nil, http.StatusOK
 }
+
+func (k *BasicAuthKeyIsValid) handleAuthFail(w http.ResponseWriter, r *http.Request, token string) (error, int) {
+
+	// Fire Authfailed Event
+	AuthFailed(k, r, token)
+
+	// Report in health check
+	reportHealthValue(k.Spec, KeyFailure, "-1")
+
+	return k.requestForBasicAuth(w, "User not authorised")
+}
+
+func (k *BasicAuthKeyIsValid) doBcryptWithCache(cacheDuration time.Duration, hashedPassword []byte, password []byte) error {
+	if err := bcrypt.CompareHashAndPassword(hashedPassword, password); err != nil {
+
+		return err
+	}
+
+	hasher := murmur3.New32()
+	hasher.Write([]byte(password))
+	hasher.Sum32()
+	k.cache.Set(string(hashedPassword), string(hasher.Sum(nil)), cacheDuration)
+
+	return nil
+}
+
+func (k *BasicAuthKeyIsValid) compareHashAndPassword(hash string, password string, logEntry *logrus.Entry) error {
+
+	cacheEnabled := !k.Spec.BasicAuth.DisableCaching
+	passwordBytes := []byte(password)
+	hashBytes := []byte(hash)
+
+	if !cacheEnabled {
+
+		logEntry.Debug("cache disabled")
+		return bcrypt.CompareHashAndPassword(hashBytes, passwordBytes)
+	}
+
+	cacheDuration := time.Duration(60) * time.Second // set a default TTL, then override based on BasicAuth.CacheTTL
+	if k.Spec.BasicAuth.CacheTTL > 0 {
+		cacheDuration = time.Duration(k.Spec.BasicAuth.CacheTTL) * time.Second
+	}
+
+	cachedPass, inCache := k.cache.Get(hash)
+	if !inCache {
+
+		logEntry.Debug("cache enabled: miss: bcrypt failover")
+		return k.doBcryptWithCache(cacheDuration, hashBytes, passwordBytes)
+	}
+
+	hasher := murmur3.New32()
+	hasher.Write([]byte(password))
+	hasher.Sum32()
+	if cachedPass.(string) != string(hasher.Sum(nil)) {
+
+		logEntry.Debug("cache enabled: hit: invalid: bcrypt failover")
+		return k.doBcryptWithCache(cacheDuration, hashBytes, passwordBytes)
```

Comment:
if it is in cache and doesn't match provided password why not to do just `bcrypt.CompareHashAndPassword(hashBytes, passwordBytes)` directly? (and eliminate one function call)  I mean there is no way `doBcryptWithCache` can put anything in cache at this point

Reply:
If user password changes, then we could get a cache hit - but an invalid password.
We call the `doBrcyptWithCache` function to re-set the cache with latest password.

Reply:
if user password changed wouldn't `compareHashAndPassword` be called with new hash and new password? I mean in case of new password `session.BasicAuthData.Password` will contain new bcrypt-hash and there is no way to get cache hit. I think new password situation won't reach this piece of logic, if I am not missing anything here.